File: 2931-maximum-spending-after-buying-items/2931-maximum-spending-after-buying-items.py
class Solution:
    def maxSpending(self, values: List[List[int]]) -> int:
        n, m = len(values), len(values[0])
        
        ptr = [m-1]*n
        h = [(values[i][-1], i) for i in range(n)]
        heapify(h)
        
        ans, d = 0, 1
        
        while h:
            price, shop = heappop(h)
            ans += price*d
            
            if(ptr[shop]>0):
                ptr[shop]-=1
                heappush(h, (values[shop][ptr[shop]], shop))
            d+=1
        
        return ans

File: 2931-maximum-spending-after-buying-items/2931-maximum-spending-after-buying-items.java
class Solution {
    public long maxSpending(int[][] values) {
        int n = values.length, m = values[0].length;
        
        int[] ptr = new int[n];
        Arrays.fill(ptr, m-1);
        
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a->a[0]));
        for(int i=0; i<n; ++i)
            pq.offer(new int[]{values[i][m-1], i});
        
        int d = 1;
        long ans = 0;
        
        while(!pq.isEmpty()){
            int price = pq.peek()[0];
            int shop = pq.peek()[1];
            pq.poll();
            
            ans += (long)price*d++;
            // ans += price*d++;    caused wrong answer
            
            if(ptr[shop] > 0){
                ptr[shop]--;
                pq.offer(new int[]{values[shop][ptr[shop]], shop});
                
            }
        }
        return ans;
    }
}